### PR TITLE
Feature/export to mp4

### PR DESCRIPTION
**Pull request by zwan0202**
## Summary

- Add export to MP4
- Refactor export functions code, and add a general export function to handle different export formats
- Link export backend to frontend

<hr/>

### What problems exist

- Progress bar is currently hardcoded
- Renderer may need to be rewritten

<hr/>

### Any sections in particular you want reviewed/discussed

Test on other OS

<hr/>

### OS

- [ ] Linux
- [ ] MacOS
- [x] Windows

### CODE REVIEW

File: src/main/export/exportToMP4.ts
Comment (chloebrett) [inline src/main/export/exportToMP4.ts]: Change return type to `Promise<void>`

File: src/main/export/exportToMP4.ts
Comment (chloebrett) [inline src/main/export/exportToMP4.ts]: Returns `Promise<void>`

File: src/main/export/exportToMP4.ts
Comment (chloebrett) [inline src/main/export/exportToMP4.ts]: `await` this

File: src/main/export/exportToMP4.ts
Comment (chloebrett) [inline src/main/export/exportToMP4.ts]: Invert this to early exit i.e.

```
if (project.transcription === null) {
  return;
}
// ... everything else
```

File: src/main/handlers/menu/setExportEnabled.ts
Comment (chloebrett) [inline src/main/handlers/menu/setExportEnabled.ts]: Lowercase acronyms

`exportEdl`
`exportMp4`

File: src/renderer/store/exportIo/actions.ts
Comment (chloebrett) [inline src/renderer/store/exportIo/actions.ts]: don't use `any` here - if the payload is empty then it's `Action<null>`

File: tools/gencode.ts
Comment (chloebrett) [inline tools/gencode.ts]: Looks like this depends on https://github.com/chloebrett/mlvet/pull/295 so wait for that to be merged first

**Review by chloebrett — CHANGES_REQUESTED** [review]
Have not yet tested functionality but code is looking pretty good

**Comment by chloebrett** [comment]
Can you add a short screen recording showing what the functionality looks like

File: tools/gencode.ts
Comment (ghost) [reply tools/gencode.ts]: https://github.com/chloebrett/mlvet/pull/295 has been merged into develop. Please merge develop into your branch. There will probably be a merge conflict with gencode as I cleaned it before merging. Please just choose develop.

File: src/main/export/exportToMP4.ts
Comment (zwan0202) [reply src/main/export/exportToMP4.ts]: I'm using `Promise.all()` here, and it returns `Promise<void[]>`

File: src/renderer/store/exportIo/actions.ts
Comment (zwan0202) [reply src/renderer/store/exportIo/actions.ts]: For some reason, if I change it to `null` the `startExport()` will not be executed, so I added `projectId` and `filePath` back

**Comment by zwan0202** [comment]
https://user-images.githubusercontent.com/37109314/191898693-17cd251f-86d6-4a99-89fd-f48be8095f81.mp4

File: src/main/export/exportToMP4.ts
Comment (chloebrett) [reply src/main/export/exportToMP4.ts]: Ah if you do `await Promise.all(…)` instead then it will return Promise<void>. You’re not using the return array so that makes more sense I feel

Comment (chloebrett) [reply src/main/export/exportToMP4.ts]: Ah if you do `await Promise.all(…)` instead then it will return `Promise<void>`. You’re not using the return array so that makes more sense I feel. Nitpick

**Comment by MutazAshhab** [comment]
Code review: :heavy_check_mark: 

Working on linux: :heavy_check_mark: 

My export testing recording: 

https://user-images.githubusercontent.com/86769131/192074770-e425ba8e-ab46-43e8-a61c-cdaefebad9ec.mp4

:warning: However, it does on linux if you do not give it a file extension. Look at this recording where it crashed because i did not add `.mp4`. Could you add a little code snippet to append the `.mp4` extension if not added by the user? 


https://user-images.githubusercontent.com/86769131/192074810-8fc75ed9-a178-45a6-bbaf-25d105491eec.mp4

**Review by MutazAshhab — APPROVED** [review]

**Review by chloebrett — APPROVED** [review]
Code looks good now, and tested locally, it works which is really cool to see - it's like a missing link in the project is now filled 😄 

Two comments
- it defaulted to 'bla.mlvet.mp4' as the file name (where bla.mlvet is the project file). This seems a bit off, should be 'bla.mp4' without the '.mlvet' in the middle
- would be really neat if we could auto-open the enclosing folder and select the exported file after the export is done - most software that exports things does this. Not sure if it would be difficult to implement. Also realise this is a product idea not an implementation critique so feel free to prioritise accordingly

**Review by exquisitebin — CHANGES_REQUESTED** [review]
Hey looking great so far. Just an issue when cancelling the export, it flashes an error window.

https://user-images.githubusercontent.com/58348527/192098257-fc35a441-51e4-42c6-98ab-a7a782a8a21f.mp4

Otherwise just finished a successful export and it works awesome 😄

Edit: I just tested this with other menu options like "Save", "Save As" and "Export to EDL" and they all have the same issue.

**Comment by zwan0202** [comment]
> Hey looking great so far. Just an issue when cancelling the export, it flashes an error window.
> 
>  2022-09-24.22-33-18_Trim.mp4 
> Otherwise just finished a successful export and it works awesome 😄
> 
> Edit: I just tested this with other menu options like "Save", "Save As" and "Export to EDL" and they all have the same issue.

I think it is because we are not handling `dialog cancel` correctly. Currently, when a dialog is canceled we just throw an error or `console.log` a message. 
Since it happens to all of them, probably fix it in a separate pr.

**Review by emmapainter — APPROVED** [review]
Looks great, thanks for getting this done!

**Review by exquisitebin — APPROVED** [review]
> > Hey looking great so far. Just an issue when cancelling the export, it flashes an error window.
> > 2022-09-24.22-33-18_Trim.mp4
> > Otherwise just finished a successful export and it works awesome 😄
> > Edit: I just tested this with other menu options like "Save", "Save As" and "Export to EDL" and they all have the same issue.
> 
> I think it is because we are not handling `dialog cancel` correctly. Currently, when a dialog is canceled we just throw an error or `console.log` a message. Since it happens to all of them, probably fix it in a separate pr.

I agree, the cancel button error should be a separate PR. Looks good to merge!

**Review by rileykeane — APPROVED** [review]

**Comment by chloebrett** [comment]
lmao you have five approvals 👏🏻 merge 👏🏻 this 👏🏻 bad 👏🏻 boy

**Pull request merged**
